### PR TITLE
only listen to subscription events upon successful `subscribe()`

### LINKED ISSUE #115: blpapi-wrapper: Properly deal with errors for all APIs
We should be careful about cleaning up state of the blpapi-wrapper Session when errors occur.

### PR DESCRIPTION
**Pull request by ericvw**
This change primarily addresses the an issue of listening to correlation
id/event names that may have failed when the underlying
`subscribe()` call is made.  Basically, listening of correlation
id/event names happens after the underlying `subscribe()` call is
successful.

The logic has also been simplified for the sake of some additional O(N)
performance.  Once the code is refactored to utilize ES6 Map, the
performance concern will eliminated.

Closes #115

### CODE REVIEW

File: lib/blpapi-wrapper.ts [hunk 1]
```diff
@@ -445,13 +421,29 @@ export class Session extends events.EventEmitter {
...
+                       util.format('Service, %s, not handled', uri));
+                var events = SERVICE_TO_SUBSCRIPTION_EVENTS_MAP[uri];
+                events.forEach((event: string): void => {
+                    log('listening on event: ' + event + ' for cid: ' + cid);
+                    this.listen(event, cid, (m: any): void => {
+                        userSubscription.emit('data', m.data, s);
+                    });
+                });
+            });
         }).nodeify(cb);
```
Comment (ericlu88) [inline lib/blpapi-wrapper.ts:453]: We'll need a `catch` block here to clean up cached `this.subscription[cid]` when `session.subscribe` throw.

**Comment by ericvw** [comment]
Updated PR with `catch`.

**Comment by ericlu88** [comment]
LGTM.

File: lib/blpapi-wrapper.ts [hunk 2]
```diff
@@ -122,15 +122,10 @@ function getServiceForSecurity(security: string): string {
...
-    subscription: Subscription;
-    serviceUri: string;
-}
-
-function getUniqueServices(sass: SubscriptionAndService[]): string[] {
-    return _(sass).map((subscriptionAndService: SubscriptionAndService): string => {
-        return subscriptionAndService.serviceUri;
-    }).uniq().valueOf();
+function subscriptionsToServices(subscriptions: Subscription[]): string[] {
+    return _.chain(subscriptions) .map( (s: Subscription): string => {
```
Comment (yupinghu) [inline lib/blpapi-wrapper.ts]: Were the spaces before `.map` and `(s: Subscription)` intentional?

**Comment by ericvw** [comment]
Addressed whitespace issue.

**Pull request merged**